### DISCUSSION TITLE
Support Vulkan 1.1 on Raspberry Pi 4

### DISCUSSION THREAD

**Discussion by mzyy94** [hecrj/iced#1278] [Ideas]
## TL;DR

I found a way to resolve `Error: GraphicsAdapterNotFound` bug on Raspberry Pi OS with **Vulkan 1.1**.

## Related issues

- https://github.com/iced-rs/iced/issues/428
- https://github.com/iced-rs/iced/pull/1160

## Description

Since [#1160](https://github.com/iced-rs/iced/pull/1160), iced on Raspberry Pi runs only on top of OpenGL with `"glow"` feature flag.

Currently Raspberry Pi 4 is Vulkan 1.1 compliant. I think _iced should work on Vulkan on Raspberry Pi 4_.

[Vulkan update: version 1.1 conformance for Raspberry Pi 4 - Raspberry Pi](https://www.raspberrypi.com/news/vulkan-update-version-1-1-conformance-for-raspberry-pi-4/)

[Install Vulkan on Raspberry Pi 4 - Q-engineering](https://qengineering.eu/install-vulkan-on-raspberry-pi.html)

I installed Vulkan 1.1 on Raspberry Pi OS (32bit) following the above instructions, and then examples of [wgpu](https://github.com/gfx-rs/wgpu) work well.

![skybox](https://user-images.githubusercontent.com/2971112/157020833-a55774e5-6978-4eb0-98d0-c751e86b8e3d.png)

<details>
<summary><code>glxinfo -B</code> output</summary>
<pre>
pi@raspberrypi:~ $ glxinfo -B
name of display: :0
display: :0  screen: 0
direct rendering: Yes
Extended renderer info (GLX_MESA_query_renderer):
    Vendor: Broadcom (0x14e4)
    Device: V3D 4.2 (0xffffffff)
    Version: 20.3.5
    Accelerated: yes
    Video memory: 3786MB
    Unified memory: yes
    Preferred profile: compat (0x2)
    Max core profile version: 0.0
    Max compat profile version: 2.1
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.1
OpenGL vendor string: Broadcom
OpenGL renderer string: V3D 4.2
OpenGL version string: 2.1 Mesa 20.3.5 (git-42a4d76fcc)
OpenGL shading language version string: 1.20

OpenGL ES profile version string: OpenGL ES 3.1 Mesa 20.3.5 (git-42a4d76fcc)
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.10

pi@raspberrypi:~ $ </pre></details>

But, running iced example that depends on wgpu renderer (such as [`svg`](https://github.com/iced-rs/iced/tree/master/examples/svg)) fails with error message `Error: GraphicsAdapterNotFound`.
It should be expected to work correctly as good as wgpu example **on Vulkan enabled platform**, but it did not.

```
pi@raspberrypi:~/Desktop/iced $ WGPU_BACKEND=vulkan cargo run --package svg
   Compiling iced_wgpu v0.4.0 (/home/pi/Desktop/iced/wgpu)
   Compiling iced v0.3.0 (/home/pi/Desktop/iced)
   Compiling svg v0.1.0 (/home/pi/Desktop/iced/examples/svg)
    Finished dev [unoptimized + debuginfo] target(s) in 1m 25s
     Running `target/debug/svg`
WARNING: v3dv is neither a complete nor a conformant Vulkan implementation. Testing use only.
Error: GraphicsAdapterNotFound
pi@raspberrypi:~/Desktop/iced $
```

I wrote [some code](https://gist.github.com/mzyy94/f67284fa865f6b4d965a4215ad788b5c) to find the bug and found that there is a problem in specifying a `limits` of [wgpu::DeviceDescriptor](https://docs.rs/wgpu/0.4.0/wgpu/struct.DeviceDescriptor.html).

```diff
diff --git a/wgpu/src/window/compositor.rs b/wgpu/src/window/compositor.rs
index 6feb795..64c5360 100644
--- a/wgpu/src/window/compositor.rs
+++ b/wgpu/src/window/compositor.rs
@@ -53,7 +53,7 @@ impl Compositor {
             .using_resolution(adapter.limits());

         #[cfg(not(target_arch = "wasm32"))]
-        let limits = wgpu::Limits::default();
+        let limits = wgpu::Limits::downlevel_defaults();

         let (device, queue) = adapter
             .request_device(
```

After changing the value of limits to `wgpu::Limits::downlevel_defaults()` shown in the patch above, the `svg` example worked good on Raspberry Pi 4.

![svg](https://user-images.githubusercontent.com/2971112/157020865-716df10e-7a42-4b30-b7b8-750f422908fc.png)

There are still some issues with Vulkan backend on Raspberry Pi 4:

- The first frame is not drawn until some events occur
- Running with `WGPU_BACKEND=gl` env-var fails
- Glitches in some projects (e.g. `game_of_life`)

![gameoflife](https://user-images.githubusercontent.com/2971112/157020946-69de3ff0-dedf-4bf9-98a6-d6ee59b24110.png)

So, this is one way to support Vulkan 1.1 on a Raspberry Pi 4. But I don't have ideas to make better patch to source code.
Any ideas?

**Comment by hecrj** [comment]
`Limits::downlevel_defaults` sounds like a better default strategy. I have applied the patch in a53fa91e0ddf374bbeb66d5e831b79127ed47a9d. Thank you! 🙇 

As to the `game_of_life` glitches, does it help if you disable the grid? What if you turn off the `antialising` field?

https://github.com/iced-rs/iced/blob/a53fa91e0ddf374bbeb66d5e831b79127ed47a9d/examples/game_of_life/src/main.rs#L23-L30

**Reply by mzyy94** [reply]
Superb patch! :tada:
The glitch continues regardless of the value of `antialising`. 😢

https://user-images.githubusercontent.com/2971112/157072117-7749d354-76de-4131-950a-1a358e5884d7.mp4
